File: apps/fago_app/lib/services/whatsapp_service.dart
import 'package:flutter/foundation.dart';
import 'package:url_launcher/url_launcher.dart';

class WhatsAppService {
  /// Opens WhatsApp natively on mobile with pre-filled message
  static Future<bool> openWhatsApp({
    required String phone,
    required String message,
  }) async {
    // Clean phone number (strip spaces, +, hyphens)
    String cleanPhone = phone.replaceAll(RegExp(r'[^\d]'), '');
    if (cleanPhone.length == 10) {
      cleanPhone = '91$cleanPhone';
    }
    final encodedMessage = Uri.encodeComponent(message);

    // Primary: Mobile App Deep-Link Intent
    final nativeAppUri = Uri.parse('whatsapp://send?phone=$cleanPhone&text=$encodedMessage');
    
    // Fallback: wa.me HTTPS Link
    final webUri = Uri.parse('https://wa.me/$cleanPhone?text=$encodedMessage');

    try {
      if (await canLaunchUrl(nativeAppUri)) {
        return await launchUrl(nativeAppUri, mode: LaunchMode.externalApplication);
      } else if (await canLaunchUrl(webUri)) {
        return await launchUrl(webUri, mode: LaunchMode.externalApplication);
      } else {
        return false;
      }
    } catch (e) {
      debugPrint('Error launching WhatsApp: $e');
      return false;
    }
  }

  /// Launch Native Google Maps App for Turn-by-Turn Navigation ($0 API Cost)
  static Future<bool> openGoogleMapsApp({
    required double destinationLat,
    required double destinationLng,
    double? originLat,
    double? originLng,
  }) async {
    // Turn-by-turn navigation intent
    final googleNavUri = Uri.parse('google.navigation:q=$destinationLat,$destinationLng&mode=d');
    
    // Web / App fallback URL
    final mapsWebUri = originLat != null && originLng != null
        ? Uri.parse('https://www.google.com/maps/dir/?api=1&origin=$originLat,$originLng&destination=$destinationLat,$destinationLng&travelmode=driving')
        : Uri.parse('https://www.google.com/maps/search/?api=1&query=$destinationLat,$destinationLng');

    try {
      if (await canLaunchUrl(googleNavUri)) {
        return await launchUrl(googleNavUri, mode: LaunchMode.externalApplication);
      } else if (await canLaunchUrl(mapsWebUri)) {
        return await launchUrl(mapsWebUri, mode: LaunchMode.externalApplication);
      } else {
        return false;
      }
    } catch (e) {
      debugPrint('Error launching Google Maps: $e');
      return false;
    }
  }

  /// Launch 1-Tap UPI Payment Intent (GPay, PhonePe, Paytm, BHIM)
  static Future<bool> openUpiPayment({
    required String upiId,
    required String name,
    required double amount,
    required String note,
  }) async {
    final encodedName = Uri.encodeComponent(name);
    final encodedNote = Uri.encodeComponent(note);
    final upiUri = Uri.parse(
      'upi://pay?pa=$upiId&pn=$encodedName&am=${amount.toStringAsFixed(2)}&cu=INR&tn=$encodedNote',
    );

    try {
      if (await canLaunchUrl(upiUri)) {
        return await launchUrl(upiUri, mode: LaunchMode.externalApplication);
      } else {
        return false;
      }
    } catch (e) {
      debugPrint('Error launching UPI Payment: $e');
      return false;
    }
  }

  /// Automated Ride Confirmation Message Template for WhatsApp
  static String getRideConfirmationTemplate({
    required String vehicleCategory,
    required String pickupAddress,
    required String dropoffAddress,
    required double fare,
  }) {
    return '🚗 *RideO Booking Request*\n\n'
        '• *Vehicle*: $vehicleCategory\n'
        '• *Pickup*: $pickupAddress\n'
        '• *Dropoff*: $dropoffAddress\n'
        '• *Estimated Fare*: ₹${fare.toStringAsFixed(0)}\n\n'
        'Please confirm availability!';
  }
}

File: apps/fago_native/app/src/main/java/com/fago/fagoapp/services/WhatsAppService.kt
package com.fago.fagoapp.services

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log

object WhatsAppService {
    private const val TAG = "WhatsAppService"

    /**
     * Opens WhatsApp natively on Android with pre-filled message.
     */
    fun openWhatsApp(context: Context, phone: String, message: String): Boolean {
        var cleanPhone = phone.replace(Regex("[^0-9]"), "")
        if (cleanPhone.length == 10) {
            cleanPhone = "91$cleanPhone"
        }
        val encodedMessage = Uri.encode(message)

        val nativeAppUri = Uri.parse("whatsapp://send?phone=$cleanPhone&text=$encodedMessage")
        val webUri = Uri.parse("https://wa.me/$cleanPhone?text=$encodedMessage")

        return try {
            val intent = Intent(Intent.ACTION_VIEW, nativeAppUri).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
            }
            context.startActivity(intent)
            true
        } catch (e: Exception) {
            Log.d(TAG, "Native WhatsApp intent failed, falling back to HTTPS wa.me: ${e.message}")
            try {
                val webIntent = Intent(Intent.ACTION_VIEW, webUri).apply {
                    flags = Intent.FLAG_ACTIVITY_NEW_TASK
                }
                context.startActivity(webIntent)
                true
            } catch (ex: Exception) {
                Log.e(TAG, "Failed to launch WhatsApp link: ${ex.message}")
                false
            }
        }
    }

    /**
     * Auto-links newly registered user's WhatsApp number to local Area Admin & WhatsApp Group.
     */
    fun autoLinkToAreaAdminWhatsAppGroup(
        context: Context,
        userPhone: String,
        userName: String,
        adminPhone: String = "919486335870",
        pincode: String = "641001"
    ) {
        val message = """
            👋 *NEW USER QR REGISTERED ON FAGO APP* 👋
            
            👤 *User Name:* $userName
            📱 *Registered WhatsApp Cell:* $userPhone
            📮 *Pincode Area:* $pincode
            
            👉 *Please add me to the official Pincode $pincode Area Admin WhatsApp Group and send my local welcome guide!*
        """.trimIndent()

        openWhatsApp(context, adminPhone, message)
    }

    /**
     * Launch Native Google Maps App for Turn-by-Turn Navigation ($0 API Cost)
     */
    fun openGoogleMapsApp(
        context: Context,
        destinationLat: Double,
        destinationLng: Double,
        originLat: Double? = null,
        originLng: Double? = null
    ): Boolean {
        val googleNavUri = Uri.parse("google.navigation:q=$destinationLat,$destinationLng&mode=d")
        val mapsWebUri = if (originLat != null && originLng != null) {
            Uri.parse("https://www.google.com/maps/dir/?api=1&origin=$originLat,$originLng&destination=$destinationLat,$destinationLng&travelmode=driving")
        } else {
            Uri.parse("https://www.google.com/maps/search/?api=1&query=$destinationLat,$destinationLng")
        }

        return try {
            val intent = Intent(Intent.ACTION_VIEW, googleNavUri).apply {
                setPackage("com.google.android.apps.maps")
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
            }
            context.startActivity(intent)
            true
        } catch (e: Exception) {
            try {
                val webIntent = Intent(Intent.ACTION_VIEW, mapsWebUri).apply {
                    flags = Intent.FLAG_ACTIVITY_NEW_TASK
                }
                context.startActivity(webIntent)
                true
            } catch (ex: Exception) {
                Log.e(TAG, "Error launching Google Maps: ${ex.message}")
                false
            }
        }
    }

    /**
     * Launch 1-Tap UPI Payment Intent (GPay, PhonePe, Paytm, BHIM)
     */
    fun openUpiPayment(
        context: Context,
        upiId: String,
        name: String,
        amount: Double,
        note: String
    ): Boolean {
        val encodedName = Uri.encode(name)
        val encodedNote = Uri.encode(note)
        val formattedAmount = String.format("%.2f", amount)
        val upiUri = Uri.parse(
            "upi://pay?pa=$upiId&pn=$encodedName&am=$formattedAmount&cu=INR&tn=$encodedNote"
        )

        return try {
            val intent = Intent(Intent.ACTION_VIEW, upiUri).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
            }
            context.startActivity(intent)
            true
        } catch (e: Exception) {
            Log.e(TAG, "Error launching UPI payment intent: ${e.message}")
            false
        }
    }

    /**
     * Automated Ride Confirmation Message Template for WhatsApp with Auto Live GPS Pin & Security OTP
     */
    fun getRideConfirmationTemplate(
        vehicleCategory: String,
        pickupAddress: String,
        dropoffAddress: String,
        fare: Double,
        pincode: String? = null,
        lat: Double? = null,
        lng: Double? = null,
        riderName: String? = null,
        riderPhone: String? = null,
        otpCode: String? = null
    ): String {
        val cleanName = if (!riderName.isNullOrEmpty() && riderName.lowercase() != "user") riderName else "FAGO Rider"
        val cleanPhone = if (!riderPhone.isNullOrEmpty()) riderPhone else "Registered Mobile"

        val sb = StringBuilder()
        sb.appendLine("🚗 *RideO Booking Request*")
        sb.appendLine("👤 *Rider*: $cleanName ($cleanPhone)")
        sb.appendLine("• *Vehicle*: $vehicleCategory")
        sb.appendLine("• *Pickup*: $pickupAddress")
        if (!pincode.isNullOrEmpty()) {
            sb.appendLine("• *Pincode*: $pincode")
        }
        sb.appendLine("• *Dropoff*: $dropoffAddress")
        sb.appendLine("• *Estimated Fare*: ₹${fare.toInt()} (0% Commission)")
        if (!otpCode.isNullOrEmpty()) {
            sb.appendLine("🔑 *START TRIP SECURITY PIN*: $otpCode")
        }

        if (lat != null && lng != null) {
            sb.appendLine("\n📍 *Live GPS Location Pin*: https://maps.google.com/?q=$lat,$lng")
        }
        sb.appendLine("\nPlease confirm availability!")
        return sb.toString()
    }
}
